### ISSUE TITLE
Allow multipart/form-data for primefaces ajax

### ISSUE THREAD

**Issue by arunvc1** [primefaces/primefaces#1355]
```
when a file upload component is present in multipart/form-data form 
<h:inputFile value="#{menuView.iconPathPart}" />      

 p:menuitem click throws the following error
```

```
27-Apr-2016 16:10:12.786 SEVERE [http-nio-8084-exec-30] null.null org.apache.tomcat.util.http.fileupload.FileUploadBase$InvalidContentTypeException: the request doesn't contain a multipart/form-data or multipart/mixed stream, content type header is application/x-www-form-urlencoded; charset=UTF-8
 javax.servlet.ServletException: org.apache.tomcat.util.http.fileupload.FileUploadBase$InvalidContentTypeException: the request doesn't contain a multipart/form-data or multipart/mixed stream, content type header is application/x-www-form-urlencoded; charset=UTF-8
    at org.apache.catalina.connector.Request.parseParts(Request.java:2732)
    at org.apache.catalina.connector.Request.getParts(Request.java:2579)
    at org.apache.catalina.connector.RequestFacade.getParts(RequestFacade.java:1083)
    at javax.servlet.http.HttpServletRequestWrapper.getParts(HttpServletRequestWrapper.java:347)
    at com.sun.faces.renderkit.html_basic.FileRenderer.decode(FileRenderer.java:91)
    at org.primefaces.renderkit.PrimeRendererWrapper.decode(PrimeRendererWrapper.java:60)
    at javax.faces.component.UIComponentBase.decode(UIComponentBase.java:832)
    at javax.faces.component.UIInput.decode(UIInput.java:771)
    at javax.faces.component.UIComponentBase.processDecodes(UIComponentBase.java:1227)
    at javax.faces.component.UIInput.processDecodes(UIInput.java:676)
    at javax.faces.component.UIComponentBase.processDecodes(UIComponentBase.java:1222)
    at javax.faces.component.UIComponentBase.processDecodes(UIComponentBase.java:1222)
    at javax.faces.component.UIComponentBase.processDecodes(UIComponentBase.java:1222)
    at javax.faces.component.UIForm.processDecodes(UIForm.java:225)
    at com.sun.faces.context.PartialViewContextImpl$PhaseAwareVisitCallback.visit(PartialViewContextImpl.java:573)
    at com.sun.faces.component.visit.PartialVisitContext.invokeVisitCallback(PartialVisitContext.java:183)
    at javax.faces.component.UIComponent.visitTree(UIComponent.java:1689)
    at javax.faces.component.UIForm.visitTree(UIForm.java:362)
    at javax.faces.component.UIComponent.visitTree(UIComponent.java:1700)
    at javax.faces.component.UIComponent.visitTree(UIComponent.java:1700)
    at com.sun.faces.context.PartialViewContextImpl.processComponents(PartialViewContextImpl.java:403)
    at com.sun.faces.context.PartialViewContextImpl.processPartial(PartialViewContextImpl.java:266)
    at org.primefaces.context.PrimePartialViewContext.processPartial(PrimePartialViewContext.java:57)
    at javax.faces.context.PartialViewContextWrapper.processPartial(PartialViewContextWrapper.java:219)
    at org.omnifaces.context.OmniPartialViewContext.processPartial(OmniPartialViewContext.java:136)
    at javax.faces.component.UIViewRoot.processDecodes(UIViewRoot.java:927)
    at com.sun.faces.lifecycle.ApplyRequestValuesPhase.execute(ApplyRequestValuesPhase.java:78)
    at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:101)
    at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:198)
    at javax.faces.webapp.FacesServlet.service(FacesServlet.java:658)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at org.omnifaces.facesviews.FacesViewsForwardingFilter.filterExtensionLess(FacesViewsForwardingFilter.java:128)
    at org.omnifaces.facesviews.FacesViewsForwardingFilter.doFilter(FacesViewsForwardingFilter.java:89)
    at org.omnifaces.filter.HttpFilter.doFilter(HttpFilter.java:108)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
    at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
    at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:610)
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:537)
    at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1085)
    at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:658)
    at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.process(Http11NioProtocol.java:222)
    at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1556)
    at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1513)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.tomcat.util.http.fileupload.FileUploadBase$InvalidContentTypeException: the request doesn't contain a multipart/form-data or multipart/mixed stream, content type header is application/x-www-form-urlencoded; charset=UTF-8
    at org.apache.tomcat.util.http.fileupload.FileUploadBase$FileItemIteratorImpl.<init>(FileUploadBase.java:800)
    at org.apache.tomcat.util.http.fileupload.FileUploadBase.getItemIterator(FileUploadBase.java:256)
    at org.apache.tomcat.util.http.fileupload.FileUploadBase.parseRequest(FileUploadBase.java:280)
    at org.apache.catalina.connector.Request.parseParts(Request.java:2668)



```

PF 6.0 or any version
JSF 2.2.12
Tomcat 8.0.15
Java 8

**Comment by tandraschko** [comment]
where is the relation to PF? Does it work without PF (e.g. with h:commandButton instead p:menuItem)?

**Comment by arunvc1** [comment]
yes it works with h:commandButton but doesn't work with p:commandButton

`<h:commandButton action="#{bean.actionMethod}" value="Submit" ><f:ajax execute="@form" render="@form" /></h:commandButton>`

Note that i am using javax.servlet.http.Part for file upload

`<h:inputFile value="#{menuView.iconPathPart}" />`

**Comment by tandraschko** [comment]
than please provide a complete example

**Comment by arunvc1** [comment]
Netbeans project with this issue is attached
[upload.zip](https://github.com/primefaces/primefaces/files/241855/upload.zip)

Run index.xhtml
Click Upoad page link
Click primefaces buttom and see the exception in console. where are h commandbutton works fine.
The same issue exist in p:menuitem

![upload1](https://cloud.githubusercontent.com/assets/16149705/14909490/d9b9ad04-0e01-11e6-9c49-78a619b316ef.png)
![upload](https://cloud.githubusercontent.com/assets/16149705/14909491/dc788a56-0e01-11e6-8528-120a173be80a.png)

**Comment by tandraschko** [comment]
Fileupload doesn't work with ajax, so if i set ajax=false on the p:commandButton, it does the same as h:commandButton.

If you don't want to upload but do some ajax actions in the same form, you should set process/execute correctl.y

**Comment by arunvc1** [comment]
It looks like a bug to me, please change both the button to ajax
h commandButton works and p commandButton doesn't work.

My point is p commandButton should do what h commandButton does, but in this case it breaks, also i am not using p fileUpload

Use the code below

```
<p:commandButton action="#{uploadView.actionUpload}" value="PF Button error" process="@form" update="@form" />
<h:commandButton action="#{uploadView.actionUpload}" value="H Button works"  ><f:ajax execute="@form" render="@form" /></h:commandButton> 
```

**Comment by tandraschko** [comment]
it just because the ajax on h:commandButton throws an JS error and therefore a full post is done.
We can't fix anything, it's expected.

**Comment by tandraschko** [comment]
but you could open a feature request for mojarra to skip the h:inputFile processing on ajax ;)

**Comment by arunvc1** [comment]
My browser shows partial/ajax see the screens

Request
![uploadpartial](https://cloud.githubusercontent.com/assets/16149705/14948960/7aa33902-1063-11e6-9776-047767abc41b.png)

Response
![uploadpartial2](https://cloud.githubusercontent.com/assets/16149705/14948956/7569229e-1063-11e6-93bd-e5af6223c2f3.png)

I think the problem is during a partial request  content-type is set to "application/x-www-form-urlencoded instead of multipart/form-data for a multipart form, setting the right content type may solve this issue.

I think Mojarra support multi part form with ajax, 
See these reference 
https://java.net/jira/browse/JAVASERVERFACES-3765
https://java.net/jira/browse/JAVASERVERFACES-3778
https://jsflive.wordpress.com/2013/04/23/jsf22-file-upload/

**Comment by tandraschko** [comment]
Feel free to provide a PR - i likely won't do it in the next months.

**Comment by arunvc1** [comment]
one easy solution that came to my mind
you don't have to implement ajax PF upload.

if form has **multipart**, is it possible to set the **content type header as multipart/form-data** on PF ajax request. That could solve this issue

Now even though the form has multipart content type header is wrongly set.
should i raise it as a different issue?

**Comment by tandraschko** [comment]
AFAICS i would just append to the file to the ajax request, than it should work.
See: http://stackoverflow.com/questions/5392344/sending-multipart-formdata-with-jquery-ajax

**Comment by arunvc1** [comment]
i think you don't have to implement to **append the file in the ajax request** just set request content type as multipart/form-data

**Comment by tandraschko** [comment]
doesn't make sense for me (it would be just an hack).
If you don't want to upload it, just don't process the inputFile and everything is fine.

**Comment by arunvc1** [comment]
1. FORM is multipart 
2. On PF button click my request content type become  application/x-www-form-urlencoded 

My point is please retail the content type header of the FORM on PF ajax button click
Hope you understand.

**Comment by melloware** [comment]
Fixed in 9.0. 

See: https://github.com/primefaces/primefaces/issues/6360
See: https://github.com/primefaces/primefaces/issues/6361
